src/Microsoft.Health.Dicom.Core/Features/Store/DicomDatasetValidator.cs

```diff
@@ -90,22 +97,55 @@ string EnsureRequiredTagIsPresent(DicomTag dicomTag)
             }
             else
             {
-                ValidateIndexedItems(dicomDataset);
+                await ValidateIndexedItems(dicomDataset, cancellationToken);
             }
         }
 
-        private void ValidateIndexedItems(DicomDataset dicomDataset)
+        private async Task ValidateIndexedItems(DicomDataset dicomDataset, CancellationToken cancellationToken)
         {
-            HashSet<DicomTag> indexableTags = QueryLimit.AllInstancesTags;
+            IReadOnlyCollection<IndexableDicomTag> indexableTags = await _indexableDicomTagService.GetIndexableDicomTagsAsync(cancellationToken);
 
-            foreach (DicomTag indexableTag in indexableTags)
+            HashSet<DicomTag> standardTags = indexableTags.Select(indexableTag => indexableTag.Tag)
+                .Where(tag => !tag.IsPrivate)
+                .ToHashSet();
+
+            IDictionary<DicomTag, DicomVR> privateTags = indexableTags.Where(indexableTag => indexableTag.Tag.IsPrivate)
+                .ToDictionary(indexableTag => indexableTag.Tag, indexableTag => indexableTag.VR);
+
+            ValidateStandardTags(dicomDataset, standardTags);
+            ValidatePrivateTags(dicomDataset, privateTags);
+        }
+
+        private void ValidateStandardTags(DicomDataset dicomDataset, HashSet<DicomTag> standardTags)
+        {
+            foreach (DicomTag indexableTag in standardTags)
             {
                 DicomElement dicomElement = dicomDataset.GetDicomItem<DicomElement>(indexableTag);
 
                 if (dicomElement != null)
                 {
-                    string value = dicomDataset.GetSingleValueOrDefault<string>(indexableTag, default);
-                    _minimumValidator.Validate(indexableTag, value);
+                    _minimumValidator.Validate(dicomElement);
+                }
+            }
+        }
+
+        private void ValidatePrivateTags(DicomDataset dicomDataset, IDictionary<DicomTag, DicomVR> privateTags)
```

Comment:
can we consolidate ValidateStandard and ValidatePrivate... Why do they need to be different?

Reply:
The way to get DicomElement for private and standard tag is different.
DicomDataset internally has a dictionary (https://github.com/fo-dicom/fo-dicom/blob/fc6347e30586feacf6152f4a38109906c93839f1/DICOM/DicomDataset.cs#L22) for all DicomItems. 
For standard tag, we can simply call DicomDataset.GetDicomItem to retrieve it from this dictionary.
For private tag, since we don't record private creator, GetDicomItem won't work, we have to loop all element inside, and locate through tag path.

Reply:
Thanks. You have DicomDataset. What not convert the CustomTagEntry to DicomTag in the Dataset which has the privateCreator and pass that to the Validate method

Reply:
will do